### PR TITLE
PF: don't unref errors when about to loop in pf_conn cb

### LINKED ISSUE #13334: concurrent_connectivity_test  tsan failure: heap_use_after_free
To reproduce:

```
tools/run_tests/run_tests.py -lc -ctsan -r"concurrent_connectivity_test" -n inf -S --force_use_pollers=epoll1
```

Note that there is more than one Tsan failures exposed by this thread and they have slightly different call-stacks. But for many of them, the top part of the call-stacks are similar.
**NOTE: Sometimes you find a tsan failure in resource-quota . That is a separate bug**

Also, note that depending on the order in which the operations happen, this is either tagged as `heap-use-after-free` or a `data-race`. Its the same issue.

```
==================
WARNING: ThreadSanitizer: heap-use-after-free (pid=139521)
  Atomic write of size 8 at 0x7d400000f600 by thread T82:
    #0 __tsan_atomic64_fetch_add <null> (concurrent_connectivity_test+0x0000004668b5)
    #1 gpr_ref /usr/local/google/home/sreek/workspace/grpc/src/core/lib/support/sync.cc:86 (concurrent_connectivity_test+0x000000578096)
    #2 grpc_error_ref /usr/local/google/home/sreek/workspace/grpc/src/core/lib/iomgr/error.cc:138 (concurrent_connectivity_test+0x0000004b1c6b)
    #3 grpc_error_create /usr/local/google/home/sreek/workspace/grpc/src/core/lib/iomgr/error.cc:355 (concurrent_connectivity_test+0x0000004b2255)
    #4 pf_connectivity_changed_locked(grpc_exec_ctx*, void*, grpc_error*) /usr/local/google/home/sreek/workspace/grpc/src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc:545 (concurrent_connectivity_test+0x00000063d675)
    #5 grpc_combiner_continue_exec_ctx /usr/local/google/home/sreek/workspace/grpc/src/core/lib/iomgr/combiner.cc:261 (concurrent_connectivity_test+0x000000520937)
    #6 grpc_exec_ctx_flush /usr/local/google/home/sreek/workspace/grpc/src/core/lib/iomgr/exec_ctx.cc:93 (concurrent_connectivity_test+0x0000004b7e72)
    #7 grpc_exec_ctx_finish /usr/local/google/home/sreek/workspace/grpc/src/core/lib/iomgr/exec_ctx.cc:56 (concurrent_connectivity_test+0x0000004b7d27)
    #8 grpc_channel_check_connectivity_state /usr/local/google/home/sreek/workspace/grpc/src/core/ext/filters/client_channel/channel_connectivity.cc:44 (concurrent_connectivity_test+0x0000004fba05)
    #9 create_loop_destroy(void*) /usr/local/google/home/sreek/workspace/grpc/test/core/surface/concurrent_connectivity_test.cc:69 (concurrent_connectivity_test+0x0000004a79e2)
    #10 thread_body(void*) /usr/local/google/home/sreek/workspace/grpc/src/core/lib/support/thd_posix.cc:42 (concurrent_connectivity_test+0x000000578f8d)

  Previous write of size 8 at 0x7d400000f600 by thread T83:
    #0 free <null> (concurrent_connectivity_test+0x000000444286)
    #1 gpr_free /usr/local/google/home/sreek/workspace/grpc/src/core/lib/support/alloc.cc:78 (concurrent_connectivity_test+0x0000005709a4)
    #2 error_destroy(grpc_error*) /usr/local/google/home/sreek/workspace/grpc/src/core/lib/iomgr/error.cc:180 (concurrent_connectivity_test+0x0000004b1e52)
    #3 grpc_error_unref /usr/local/google/home/sreek/workspace/grpc/src/core/lib/iomgr/error.cc:192 (concurrent_connectivity_test+0x0000004b1d95)
    #4 unref_errs(grpc_error*) /usr/local/google/home/sreek/workspace/grpc/src/core/lib/iomgr/error.cc:153 (concurrent_connectivity_test+0x0000004b4a44)
    #5 error_destroy(grpc_error*) /usr/local/google/home/sreek/workspace/grpc/src/core/lib/iomgr/error.cc:177 (concurrent_connectivity_test+0x0000004b1e1e)
    #6 grpc_error_unref /usr/local/google/home/sreek/workspace/grpc/src/core/lib/iomgr/error.cc:192 (concurrent_connectivity_test+0x0000004b1d95)
    #7 grpc_connectivity_state_set /usr/local/google/home/sreek/workspace/grpc/src/core/lib/transport/connectivity_state.cc:185 (concurrent_connectivity_test+0x00000055c334)
    #8 set_channel_connectivity_state_locked(grpc_exec_ctx*, client_channel_channel_data*, grpc_connectivity_state, grpc_error*, char const*) /usr/local/google/home/sreek/workspace/grpc/src/core/ext/filters/client_channel/client_channel.cc:255 (concurrent_connectivity_test+0x0000005038b0)
    #9 start_transport_op_locked(grpc_exec_ctx*, void*, grpc_error*) /usr/local/google/home/sreek/workspace/grpc/src/core/ext/filters/client_channel/client_channel.cc:624 (concurrent_connectivity_test+0x00000050338f)
    #10 grpc_combiner_continue_exec_ctx /usr/local/google/home/sreek/workspace/grpc/src/core/lib/iomgr/combiner.cc:261 (concurrent_connectivity_test+0x000000520937)
    #11 grpc_exec_ctx_flush /usr/local/google/home/sreek/workspace/grpc/src/core/lib/iomgr/exec_ctx.cc:93 (concurrent_connectivity_test+0x0000004b7e72)
    #12 grpc_exec_ctx_finish /usr/local/google/home/sreek/workspace/grpc/src/core/lib/iomgr/exec_ctx.cc:56 (concurrent_connectivity_test+0x0000004b7d27)
    #13 grpc_channel_destroy /usr/local/google/home/sreek/workspace/grpc/src/core/lib/surface/channel.cc:427 (concurrent_connectivity_test+0x0000004e1d35)
    #14 create_loop_destroy(void*) /usr/local/google/home/sreek/workspace/grpc/test/core/surface/concurrent_connectivity_test.cc:78 (concurrent_connectivity_test+0x0000004a7b48)
    #15 thread_body(void*) /usr/local/google/home/sreek/workspace/grpc/src/core/lib/support/thd_posix.cc:42 (concurrent_connectivity_test+0x000000578f8d)

  Thread T82 (tid=141222, running) created by main thread at:
    #0 pthread_create <null> (concurrent_connectivity_test+0x000000448853)
    #1 gpr_thd_new /usr/local/google/home/sreek/workspace/grpc/src/core/lib/support/thd_posix.cc:66 (concurrent_connectivity_test+0x000000578e73)
    #2 run_concurrent_connectivity_test() /usr/local/google/home/sreek/workspace/grpc/test/core/surface/concurrent_connectivity_test.cc:221 (concurrent_connectivity_test+0x0000004a8730)
    #3 main /usr/local/google/home/sreek/workspace/grpc/test/core/surface/concurrent_connectivity_test.cc:291 (concurrent_connectivity_test+0x0000004a8d69)

  Thread T83 (tid=141242, finished) created by main thread at:
    #0 pthread_create <null> (concurrent_connectivity_test+0x000000448853)
    #1 gpr_thd_new /usr/local/google/home/sreek/workspace/grpc/src/core/lib/support/thd_posix.cc:66 (concurrent_connectivity_test+0x000000578e73)
    #2 run_concurrent_connectivity_test() /usr/local/google/home/sreek/workspace/grpc/test/core/surface/concurrent_connectivity_test.cc:221 (concurrent_connectivity_test+0x0000004a8730)
    #3 main /usr/local/google/home/sreek/workspace/grpc/test/core/surface/concurrent_connectivity_test.cc:291 (concurrent_connectivity_test+0x0000004a8d69)

SUMMARY: ThreadSanitizer: heap-use-after-free (/usr/local/google/home/sreek/workspace/grpc/bins/tsan/concurrent_connectivity_test+0x4668b5) in __tsan_atomic64_fetch_add
==================
FAILED: bins/tsan/concurrent_connectivity_test GRPC_POLL_STRATEGY=epoll1 [ret=66, pid=139521, time=13.4sec]
FLAKE: bins/tsan/concurrent_connectivity_test GRPC_POLL_STRATEGY=epoll1 [1/84 runs flaked]
```

### PR DESCRIPTION
**Pull request by dgquintas**
If a new iteration awaits, don't unref the newly retrieved error instance. 

Fixes #13334

---
This change is [Reviewable](https://reviewable.io/reviews/grpc/grpc/13343)

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                              FILE SIZE
 ++++++++++++++ GROWING                                                                ++++++++++++++
  +0.7%     +32 src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc     +32  +0.7%
      +1.7%     +28 pf_connectivity_changed_locked                                             +28  +1.7%
      +3.2%      +4 [Unmapped]                                                                  +4  +3.2%

 -------------- SHRINKING                                                              --------------
  [ = ]       0 [None]                                                                     -16  -0.0%

  +0.0%     +32 TOTAL                                                                      +16  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                              FILE SIZE
 ++++++++++++++ GROWING                                                                ++++++++++++++
  +1.7%     +80 src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc     +80  +1.7%
      +5.3%     +88 pf_connectivity_changed_locked                                             +88  +5.3%

 -+-+-+-+-+-+-+ MIXED                                                                  +-+-+-+-+-+-+-
  -0.0%     -16 [None]                                                                    +320  +0.0%

  +0.0%     +64 TOTAL                                                                     +400  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

File: src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc [hunk 1]
```diff
@@ -506,11 +508,13 @@ static void pf_connectivity_changed_locked(grpc_exec_ctx* exec_ctx, void* arg,
         }
         sd->curr_connectivity_state =
```
Comment (markdroth) [inline src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc]: What happens if we simply remove the call to `grpc_subchannel_check_connectivity()` and instead unconditionally call `grpc_lb_subchannel_data_start_connectivity_watch()`?  Does anything break?

File: src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc [hunk 2]
```diff
@@ -553,11 +559,13 @@ static void pf_connectivity_changed_locked(grpc_exec_ctx* exec_ctx, void* arg,
...
```
Comment (markdroth) [inline src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc]: Same here.

File: src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc (hunk 1 above)
Comment (dgquintas) [reply src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc]: Excellent suggestion. I've also removed that pesky loop around the switch.

File: src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc (hunk 2 above)
Comment (dgquintas) [reply src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc]: Ditto.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                              FILE SIZE
 ++++++++++++++ GROWING                                                                ++++++++++++++

 -------------- SHRINKING                                                              --------------
  -3.0%    -144 src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc    -144  -3.0%
      -8.7%    -146 pf_connectivity_changed_locked                                            -146  -8.7%
  -0.0%    -112 [None]                                                                     -80  -0.0%
      -0.0%     -80 [Unmapped]                                                                 -80  -0.0%
      -4.0%     -32 [None]                                                                       0  [ = ]

  -0.0%    -256 TOTAL                                                                     -224  -0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by dgquintas** [comment]
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc, line 445 at r2](https://reviewable.io:443/reviews/grpc/grpc/13343#-KybGK5AMRgEDzPBJXJa:-KybOkt5tAZYUyo7_pF1:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/b07d7a8051eaa7aef58cd45733b54f072a9e80f3/src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc#L445)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Now that we're not looping, I think we can move this down into the cases for TRANSIENT_FAILURE and SHUTDOWN below, rather than special-casing them here.
</blockquote></details>

Done.

---

*[src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc, line 447 at r2](https://reviewable.io:443/reviews/grpc/grpc/13343#-KybFq82nsVMaJ6JgXrR:-KybOlC5CLXZslZtSgf2:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/b07d7a8051eaa7aef58cd45733b54f072a9e80f3/src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc#L447)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

I don't think we actually need this anymore, because we've eliminated all the cases where it was set to true.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/grpc/grpc/13343)*

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                              FILE SIZE
 ++++++++++++++ GROWING                                                                ++++++++++++++

 -------------- SHRINKING                                                              --------------
  -3.4%    -160 src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc    -160  -3.4%
      -9.2%    -153 pf_connectivity_changed_locked                                            -153  -9.2%
      -5.6%      -7 [Unmapped]                                                                  -7  -5.6%
  -0.0%    -128 [None]                                                                    -136  -0.0%
      -0.0%     -96 [Unmapped]                                                                -136  -0.0%
      -4.0%     -32 [None]                                                                       0  [ = ]

  -0.0%    -288 TOTAL                                                                     -296  -0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by dgquintas** [comment]
Issues: #13122 #11008 #13330

**Pull request merged**